panelapp/api/v1/viewsets.py

```diff
@@ -197,6 +204,123 @@ class EntityViewSet(viewsets.mixins.ListModelMixin, viewsets.GenericViewSet):
+        tags = self.request.query_params.get("tags")
+
+        if entity_name and confidence_level and tags:
+            if (
+                object["entity_name"] in entity_name.split(",")
+                and object["confidence_level"] == confidence_level
+                and [True for tag in tags.split(",") if tag in object["tags"]]
+            ):
+                return True
+            else:
+                return False
+        elif entity_name and confidence_level:
+            if (
+                object["entity_name"] in entity_name.split(",")
+                and object["confidence_level"] == confidence_level
+            ):
+                return True
+            else:
+                return False
+        elif entity_name and tags:
+            if object["entity_name"] in entity_name.split(",") and [
+                True for tag in tags.split(",") if tag in object["tags"]
+            ]:
+                return True
+            else:
+                return False
+        elif confidence_level and tags:
+            if object["confidence_level"] == confidence_level and [
+                True for tag in tags.split(",") if tag in object["tags"]
+            ]:
+                return True
+            else:
+                return False
+        elif entity_name:
+            if object["entity_name"] in entity_name.split(","):
+                return True
+            else:
+                return False
+        elif confidence_level:
+            if object["confidence_level"] == confidence_level:
+                return True
+            else:
+                return False
+        elif tags:
+            for tag in tags.split(","):
+                if tag in object["tags"]:
+                    return True
+            else:
+                return False
+        else:
+            return True
+
+    def list(self, request, *args, **kwargs):
+        version = self.request.query_params.get("version", None)
+        if version:
+            try:
+                major_version, minor_version = version.split(".")
+            except ValueError:
+                raise APIException(
+                    detail="Incorrect version supplied", code="incorrect_version"
+                )
+
+            obj = HistoricalSnapshot.objects.filter(
+                panel__pk=self.kwargs["panel_pk"],
+                major_version=major_version,
+                minor_version=minor_version,
+            ).first()
+
+            if obj:
+                count = len(obj.data["genes"])
+                start = 0
+                finish = 100
+                page = self.request.query_params.get("page", None)
+                response = {
+                    "count": count,
+                    "next": None,
+                    "previous": None,
+                    "results": [],
+                }
+                max_pages = ceil(count / 100)
+
+                if max_pages > 1:
```

Comment:
@Bash-Maistora could you move pagination to a separate method, so it's easy to test?
Btw, I don't see tests for historical data pagination and filtering - could you add it as well?